iceoryx_dds/include/iceoryx_dds/internal/gateway/dds_gateway_generic.inl

```diff
@@ -0,0 +1,160 @@
+#include "iceoryx_dds/dds/dds_types.hpp"
+#include "iceoryx_dds/internal/log/logging.hpp"
+#include <iceoryx_utils/internal/file_reader/file_reader.hpp>
+
+
+// ================================================== Public ================================================== //
+
+template <typename channel_t, typename gateway_t>
+inline iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::~DDSGatewayGeneric() noexcept
+{
+    shutdown();
+}
+
+template <typename channel_t, typename gateway_t>
+inline void iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::runMultithreaded() noexcept
+{
+    m_discoveryThread = std::thread([this] { discoveryLoop(); });
+    m_forwardingThread = std::thread([this] { forwardingLoop(); });
+    m_isRunning.store(true, std::memory_order_relaxed);
+}
+
+template <typename channel_t, typename gateway_t>
+inline void iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::shutdown() noexcept
+{
+    if (m_isRunning.load(std::memory_order_relaxed))
+    {
+        iox::dds::LogDebug() << "[DDSGatewayGeneric] Shutting down Posh2DDSGateway.";
+
+        m_isRunning.store(false, std::memory_order_relaxed);
+
+        m_discoveryThread.join();
+        m_forwardingThread.join();
+    }
+}
+
+template <typename channel_t, typename gateway_t>
+inline uint64_t iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::getNumberOfChannels() const noexcept
+{
+    auto guardedVector = m_channels.GetScopeGuard();
+    return guardedVector->size();
+}
+
+// ================================================== Protected ================================================== //
+
+template <typename channel_t, typename gateway_t>
+inline iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::DDSGatewayGeneric() noexcept
+    : gateway_t(iox::capro::Interfaces::DDS)
+{
+    LogDebug() << "[DDSGatewayGeneric] Using default channel factory.";
+    m_channelFactory = channel_t::create;
+}
+
+template <typename channel_t, typename gateway_t>
+inline channel_t
+iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::addChannel(const iox::capro::ServiceDescription& service) noexcept
+{
+    auto channel = m_channelFactory(service);
+    m_channels->push_back(channel);
+    iox::dds::LogDebug() << "[DDSGatewayGeneric] Channel set up for service: "
+                         << "/" << service.getInstanceIDString() << "/" << service.getServiceIDString() << "/"
+                         << service.getEventIDString();
+    return channel;
+}
+
+template <typename channel_t, typename gateway_t>
+inline iox::cxx::optional<channel_t>
+iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::findChannel(const iox::capro::ServiceDescription& service) const
+    noexcept
+{
+    auto guardedVector = this->m_channels.GetScopeGuard();
+    auto channel = std::find_if(guardedVector->begin(), guardedVector->end(), [&service](const channel_t& channel) {
+        return channel.getService() == service;
+    });
+    if (channel == guardedVector->end())
+    {
+        return iox::cxx::nullopt_t();
```

Comment:
just a hint, in master there should now be a `iox::cxx::nullopt` constexpr. So you can now do a `return cxx::nullopt;`

Reply:
It didn't work. I didn't spend time trying to make it work though.